Automattic/woocommerce-payments
# Support new platform checkout user from checkout blocks

Fixes #4614 

### Description
Added support to create new platform checkout user from Checkout Blocks.
- Inserting the existing `CheckoutPageSaveUser` in blocks checkout page if platform checkout is enabled.
- On blocks checkout page the status of `save user` checkbox and the phone number do not make their way to the backend on POST request. On shortcode, adding a proper name to the element was enough to piggyback them with the POST request but blocks only sends [this set of data](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/base/context/providers/cart-checkout/checkout-processor.js#L185-L193) in request body. To get the values in backend used [extensionCartUpdate](https://github.com/woocommerce/woocommerce-blocks/blob/00a1174aebefae2b37178a3ded68cc6c8d56bbf4/docs/third-party-developers/extensibility/rest-api/extend-rest-api-update-cart.md#registering-a-callback-to-run-when-the-cartextensions-endpoint-is-hit) as the solution.
- As part of the above solution, registered a callback for `cart/extensions` endpoint and calling the `extensionCartUpdate` function with desired data when the `Place Order` button is clicked.
- As we need the value of `save_user_in_platform_checkout` selection in multiple places in backend (basically in multiple steps in a checkout flow), saved the value in `WC` session.

### Testing steps 
- Use `[shortcode-checkout]` in the checkout page. Ensure that nothing is broken there, new and existing user flow works same as before.
- Use `Checkout` block in the checkout page
- Go to the store, add some product and go to the blocks checkout page.
- Adding a new platform checkout user email in the email field will show the `Remember your details` section. Select the `save user` checkbox, complete the checkout with this new email. This should create a new platform checkout user.
- Adding an existing platform checkout user in the email field will hide the `Remember your details` section.
- Check and ensure that the experience on the blocks checkout page exactly same as the existing feature in the short-code checkout page.
